src/client_side.cc

```diff
@@ -1148,26 +1148,28 @@ prepareAcceleratedURL(ConnStateData * conn, const Http1::RequestParserPointer &h
     if (vport < 0)
         vport = conn->clientConnection->local.port();
 
-    char *host = NULL;
-    if (vhost && (host = hp->getHostHeaderField())) {
+    char *receivedHost = nullptr;
+    if (vhost && (receivedHost = hp->getHostHeaderField())) {
+        SBuf host(receivedHost);
         debugs(33, 5, "ACCEL VHOST REWRITE: vhost=" << host << " + vport=" << vport);
-        char thost[256];
         if (vport > 0) {
-            thost[0] = '\0';
-            char *t = NULL;
-            if (host[strlen(host) - 1] != ']' && (t = strrchr(host,':')) != nullptr) {
-                strncpy(thost, host, (t-host));
-                snprintf(thost+(t-host), sizeof(thost)-(t-host), ":%d", vport);
-                host = thost;
-            } else if (!t) {
-                snprintf(thost, sizeof(thost), "%s:%d",host, vport);
-                host = thost;
+            const auto lastColonPos = host.rfind(':');
+            if (lastColonPos != SBuf::npos) {
+                if (*host.rbegin() != ']') {
+                    // address with a port (IPv4, IPv6, or domain name)
+                    host.chop(0, lastColonPos); // truncate until the last colon
+                    host.appendf(":%d", vport);
+                }
+                // else: IPv6 address without a port
```

Comment:
Please check whether this case needs to be handled (differently).

Reply:
good catch. yes the appendf line should be out a few scopes. That makes the new code much simpler.

Reply:
I have fixed this in 0a4a970